design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
+
+Software RAID is constrained in other ways:
+
+- There must be 0, 1, or 2 software RAID devices.
+- The first software RAID volume must use RAID level 1 to ensure the
+  host boots in the event of a disk failure.
+- The second software RAID volume may use RAID level `0`, `1`, or
+  `1+0`.
+
+Regardless of the type of RAID used, the first volume defined is used
+as the root volume.
+
+The status fields of the `BareMetalHost` expand to include the RAID
+configuration used for the most recent provisioning operation.
+
+```yaml
+status:
+  provisioning:
+    raid:
+      hardwareVolumes: ...
+      softwareVolumes: ...
+```
+
+### Controller Changes
+
+Ironic applies RAID configuration during a cleaning phase, separate
+from the provisioning phase. In order for the metal3 controller to
+track this new phase, a new "preparing" state is introduced to the
```

Comment:
Yep, exactly this thing is not needed if we use deploy steps.

Reply:
I believe @dtantsur is referring to his proposal #138 to add sub-states within Provisioning. This would allow us to track the progress of cleaning without adding a new provisioning state, and would give more accurate information to the user about what's actually happening. https://github.com/metal3-io/metal3-docs/pull/138#discussion_r504718910 (e.g. we don't actually wait for cleaning to finish before moving from Preparing->Provisioning, so it's actually quite misleading if you care about low-level details for debugging purposes.)
There's nothing to implement in gophercloud, we just need to decide if this is the path forward and add the new sub-state to the CRD.
If we add the Preparing state first then subsequently switching to sub-states is a lot more painful than doing it from the outset.